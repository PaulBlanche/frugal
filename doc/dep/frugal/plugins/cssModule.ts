export * from "frugal/plugins/cssModule.ts";
