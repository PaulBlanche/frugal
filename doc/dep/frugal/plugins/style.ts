export * from 'frugal/plugins/style.ts';
