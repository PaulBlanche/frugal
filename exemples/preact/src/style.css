html,
body {
    font-family: system-ui;
    margin: 0;
}

body {
    padding: 2rem;
}

main {
    max-width: 600px;
    margin: 0 auto;
}

.scroll-padder {
    padding: 200px 0 400px;
    text-align: center;
}