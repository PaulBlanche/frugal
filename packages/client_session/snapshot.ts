import * as murmur from '../murmur/mod.ts';

type Node = {
    name: string;
    attributes: { name: string; value: string }[];
    content: string;
    hash: string;
    element: Element;
};

export type Head = Node[];

export function serialize(head: HTMLHeadElement) {
    const snapshot: Head = [];

    head.childNodes.forEach((node) => {
        if (!(node instanceof Element)) {
            return;
        }

        const hash = new murmur.Hash();

        const name = node.nodeName.toLowerCase();
        const content = node.textContent ?? '';

        const sorted: string[] = [];
        const attributes: { name: string; value: string }[] = [];
        for (let i = 0, length = node.attributes.length; i < length; i++) {
            const attribute = node.attributes[i];
            attributes.push({
                name: attribute.name,
                value: attribute.value,
            });
            sorted.push(`${attribute.name}="${attribute.value}"`);
        }

        hash.update(name);

        sorted.sort();
        sorted.forEach((value) => {
            hash.update(value);
        });

        hash.update(content);

        snapshot.push({
            name,
            attributes,
            content,
            hash: hash.digest(),
            element: node,
        });
    });

    return snapshot;
}
