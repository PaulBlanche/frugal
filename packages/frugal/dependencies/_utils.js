/**
 * @template T
 * @param {AsyncIterable<T>} stream
 * @param {number} interval
 */
export function* debounce(stream, interval) {
    /** @type {boolean | undefined} */
    let first; // is this first event?  will pass
    /** @type {T[]} */
    let buffer = []; // the last event raised
    /** @type {Promise<T[]>} */
    let deferred; // deferred promise instance
    /** @type {(value: T[]) => void} */
    let resolve; // resolve method for deferred promise

    /** @param {boolean} isFirst */
    function reset(isFirst) {
        first = isFirst;
        buffer.length = 0;
        deferred = new Promise((r) => (resolve = r));
    }

    function passEvent() {
        // if no event to pass
        if (buffer.length === 0) {
            first = true; // reset first state
            return;
        }

        const eventsToEmit = [...buffer];
        const res = resolve;
        reset(false);
        setTimeout(passEvent, interval);
        res(eventsToEmit);
    }

    reset(true);
    destreamify(stream, (event) => {
        buffer.push(event);
        if (first) {
            passEvent();
        }
    });

    while (true) {
        // @ts-expect-error: deferred is definitly assigned before in `reset` call
        yield deferred;
    }
}

/**
 * @template T
 * @param {AsyncIterable<T>} stream
 * @param {(event: T) => void} callback
 */
async function destreamify(stream, callback) {
    for await (let event of stream) {
        callback(event);
    }
}

/**
 * @template T
 * @param {AsyncIterable<T>[]} iterable
 * @returns {AsyncIterable<T>}
 */
export async function* combine(iterable) {
    const asyncIterators = Array.from(iterable, (o) => o[Symbol.asyncIterator]());
    const results = [];
    let count = asyncIterators.length;
    const never = new Promise(() => {});

    /**
     * @param {AsyncIterator<T>} asyncIterator
     * @param {number} index
     * @returns
     */
    function getNext(asyncIterator, index) {
        return asyncIterator.next().then((result) => ({
            index,
            result,
        }));
    }

    const nextPromises = asyncIterators.map(getNext);

    try {
        while (count) {
            const { index, result } = await Promise.race(nextPromises);
            if (result.done) {
                nextPromises[index] = never;
                results[index] = result.value;
                count--;
            } else {
                nextPromises[index] = getNext(asyncIterators[index], index);
                yield result.value;
            }
        }
    } finally {
        for (const [index, iterator] of asyncIterators.entries())
            if (nextPromises[index] != never && iterator.return != null) {
                iterator.return();
            }
    }
    return results;
}
