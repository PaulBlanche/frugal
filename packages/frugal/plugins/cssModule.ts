export * from "../src/plugins/cssModule/cssModule.js";
