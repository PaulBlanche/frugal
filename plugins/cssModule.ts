export { cssModule } from '../src/build/plugins/cssModule/mod.ts';
