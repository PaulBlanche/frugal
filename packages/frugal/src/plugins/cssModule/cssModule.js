import * as _type from "./_type/cssModule.js";
export * from "./_type/cssModule.js";

import * as path from "../../../dependencies/path.js";
import * as fs from "../../../dependencies/fs.js";
import * as xxhash from "../../../dependencies/xxhash.js";

import * as plugin from "../../Plugin.js";
import { Bundler } from "./Bundler.js";

/**
 * @param {Partial<_type.CssModuleOptions>} [param0]
 * @returns {plugin.Plugin}
 */
export function cssModule({ filter = /\.module.css$/, pattern, dashedIdents } = {}) {
    return (frugal) => {
        return {
            name: "frugal:cssModule",
            setup(build) {
                const cssLoader = build.initialOptions.loader?.[".css"] ?? "css";

                const bundler = new Bundler({
                    sourceMap: Boolean(build.initialOptions.sourcemap),
                    projectRoot: build.initialOptions.absWorkingDir,
                    options: {
                        pattern,
                        dashedIdents,
                    },
                });

                /** @type {Map<string, Uint8Array>} */
                const cssCache = new Map();

                build.onResolve({ filter: /\.frugal-compiled-css-module\.css$/ }, (args) => {
                    if (cssCache.has(args.path)) {
                        return { path: args.path, namespace: "virtual" };
                    }
                });

                build.onResolve({ filter: /cssModuleHelper:format\.js/ }, (args) => {
                    return { path: args.path, namespace: "cssModuleHelper" };
                });

                build.onLoad(
                    { filter: /cssModuleHelper:format\.js/, namespace: "cssModuleHelper" },
                    async (args) => {
                        return {
                            loader: "js",
                            contents: await fs.readFile(
                                path.fromFileURL(import.meta.resolve("./format.js")),
                            ),
                        };
                    },
                );

                build.onResolve({ filter }, (args) => {
                    return {
                        path: path.resolve(path.dirname(args.importer), args.path),
                        namespace: args.namespace,
                    };
                });

                build.onLoad(
                    { filter: /\.frugal-compiled-css-module\.css$/, namespace: "virtual" },
                    (args) => {
                        const contents = cssCache.get(args.path);
                        return { loader: cssLoader, contents, resolveDir: path.dirname(args.path) };
                    },
                );

                build.onLoad({ filter }, async (args) => {
                    const contents = await fs.readFile(args.path);
                    const contentHash = (await xxhash.create()).update(contents).digest();

                    const cssPath = path.resolve(
                        path.dirname(args.path),
                        `${path.basename(args.path)}-${contentHash}.frugal-compiled-css-module.css`,
                    );

                    const module = await bundler.bundle(args.path, cssPath, contents);

                    cssCache.set(cssPath, module.css);

                    return { loader: "js", contents: module.js };
                });
            },
        };
    };
}
