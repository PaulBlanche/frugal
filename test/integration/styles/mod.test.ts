import { style } from '../../../plugins/style.ts';
import { cssModule } from '../../../plugins/cssModule.ts';
import { assertSnapshot } from '../../../dep/std/testing/snapshot.ts';
import { getBuilder, withBrowser } from '../../_utils.ts';
import * as asserts from '../../../dep/std/testing/asserts.ts';

Deno.test('script_loader: file structure', async (t) => {
    const builder = getBuilder({
        self: import.meta.url,
        pages: ['./page-bar.ts', './page-foo.ts'],
        log: {
            level: 'silent',
        },
        plugins: [
            cssModule(),
            style(),
        ],
    });

    const frugal = await builder.build();

    const responseCache = await frugal.config.responseCache('runtime');
    const pageFoo1 = (await responseCache.get('/foo/1'))?.body!;
    const pageFoo2 = (await responseCache.get('/foo/2'))?.body!;
    const pageBar1 = (await responseCache.get('/bar/1'))?.body!;
    const pageBar2 = (await responseCache.get('/bar/2'))?.body!;

    await assertSnapshot(t, pageFoo1);
    await assertSnapshot(t, pageFoo2);
    await assertSnapshot(t, pageBar1);
    await assertSnapshot(t, pageBar2);

    await builder.clean();
});

Deno.test('script_loader: style order', async (t) => {
    const builder = getBuilder({
        self: import.meta.url,
        pages: ['./page-bar.ts', './page-foo.ts'],
        plugins: [
            cssModule(),
            style(),
        ],
    });

    const frugal = await builder.build();
    const controller = new AbortController();
    frugal.serve({ signal: controller.signal });

    await withBrowser(async (browser) => {
        const page = await browser.newPage();

        await page.goto('http://localhost:8000/foo/1');

        const fooStyles = await page.evaluate(getStyles);

        asserts.assertEquals(fooStyles, {
            page: ['rgb(128, 0, 128)', 'rgb(255, 0, 0)'],
            shared: ['rgb(0, 255, 255)'],
            global: ['rgb(255, 255, 0)'],
        });

        await page.goto('http://localhost:8000/bar/1');

        const barStyles = await page.evaluate(getStyles);

        asserts.assertEquals(barStyles, {
            page: ['rgb(128, 0, 128)', 'rgb(0, 0, 255)'],
            shared: ['rgb(0, 255, 255)'],
            global: ['rgb(255, 255, 0)'],
        });

        function getStyles() {
            const pageElement = document.getElementById('page');
            const sharedElement = document.getElementById('shared');
            const globalElement = document.getElementById('global');
            const pageStyle = getComputedStyle(pageElement!);
            const sharedStyle = getComputedStyle(sharedElement!);
            const globalStyle = getComputedStyle(globalElement!);
            return {
                page: [pageStyle.color, pageStyle.backgroundColor],
                shared: [sharedStyle.color],
                global: [globalStyle.color],
            };
        }
    }, {
        onClose: () => controller.abort(),
    });

    await builder.clean();
});
