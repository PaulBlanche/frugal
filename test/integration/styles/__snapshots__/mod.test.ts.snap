export const snapshot = {};

snapshot[`script_loader: file structure 1`] = `
'<html>
    <head>
        <title>bar</title>
        <link rel="stylesheet" href="/css/page-foo.css" />
    </head>
    <body>
        <div id="page" class="dA6UxW_foo oUtnuW_common">page</div>
        <div id="shared" class="IJOK1q_shared">shared</div>
        <div id="global" class="global">global</div>
    </body>
</html>'
`;

snapshot[`script_loader: file structure 2`] = `
'<html>
    <head>
        <title>bar</title>
        <link rel="stylesheet" href="/css/page-foo.css" />
    </head>
    <body>
        <div id="page" class="dA6UxW_foo oUtnuW_common">page</div>
        <div id="shared" class="IJOK1q_shared">shared</div>
        <div id="global" class="global">global</div>
    </body>
</html>'
`;

snapshot[`script_loader: file structure 3`] = `
'<html>
    <head>
        <title>bar</title>
        <link rel="stylesheet" href="/css/page-bar.css" />
    </head>
    <body>
        <div id="page" class="S4Nd0G_bar oUtnuW_common">page</div>
        <div id="shared" class="IJOK1q_shared">shared</div>
        <div id="global" class="global">global</div>
    </body>
</html>'
`;

snapshot[`script_loader: file structure 4`] = `
'<html>
    <head>
        <title>bar</title>
        <link rel="stylesheet" href="/css/page-bar.css" />
    </head>
    <body>
        <div id="page" class="S4Nd0G_bar oUtnuW_common">page</div>
        <div id="shared" class="IJOK1q_shared">shared</div>
        <div id="global" class="global">global</div>
    </body>
</html>'
`;
